Reevaluate task prerequisites if they change.

damphyr opened the pull request:
This allows a prerequisite to add additional prereqs to it's parent task.

For a bit more blurb see http://ampelofilosofies.gr/software/2012/01/13/rake-dynamic-prerequisites/

This will cut my current rakefile size down by a good 1/4. I'll have to run some benchmarks to see how much faster the startup times get.

damphyr committed:
reevaluate prerequisites if they change. This allows a prerequisite to add additional prereqs to it's parent task

damphyr committed:
allow multi_task to use dynamic prereqs as well

damphyr committed:
fixed copy-paste error, added the test that caught it

zenspider commented on lib/rake/task.rb:
```diff
@@ -171,10 +171,12 @@ def add_chain_to(exception, new_chain)
         prereq_args = task_args.new_scope(prereq.arg_names)
         prereq.invoke_with_call_chain(prereq_args, invocation_chain)
       }
+      invoke_prerequisites(task_args, invocation_chain) if prerequisite_tasks.size != original_size
```
This can trivially be rewritten to not be recursive and still be just as clear.

damphyr replied:
You mean by duping the prerequisites array and then iterating over the difference?

orig = prerequisite_tasks.dup
prerequisite_tasks.each{}
#iterate over any additional prereqs
(prerequisite_tasks-orig).each

it comes out as the same LOCs and does less iterations. The additional memory should not be an issue I think.

damphyr committed:
reevaluate prerequisites without resorting to recursion

Muriel-Salvan commented:
What if prerequisites are changed again during the re-evaluation ?

Maybe this should be better in this case ?

invoked_prerequisites = []
while !(prerequisites_to_invoke = (prerequisite_tasks - invoked_prerequisites)).empty?
  invoke_prerequisite_list(prerequisites_to_invoke, args, invocation_chain)
  invoked_prerequisites.concat(prerequisites_to_invoke)
end

damphyr commented:
Practically it makes little sense (we are consciously creating these task dependencies after all) but the code wouldn't be technically correct if it didn't take it into account.

damphyr committed:
process the prerequisite list dynamically

A task now will run through the prereqs list for as long as the list
changes as Muriel-Salvan suggested.
It's also non-recursive and the case where subsequent dependencies
enhance the prerequisite list is also tested

damphyr commented:
As an aside, I did a "rake test" with the latest jruby by mistake and got failures. It seems that the stock 0.9.2.2 is always ahead in the require path in jruby so the code in my repository was not loaded. Running the unit tests with "ruby -I lib/ "
picks the correct sources and the tests go green in jruby as well.

jimweirich commented:
First, sorry for taking so long to respond.  I'm trying to go through the backlog and clear out requests.

If I understand your argument in http://ampelofilosofies.gr/software/2012/01/13/rake-dynamic-prerequisites/, you are saying that

```
task :build => :determine_dependencies
```

is faster than

```
task :build => determine_dependencies()
```

because the dependencies are scanned once when the determine_dependencies task is invoked rather then each time the determine_dependencies() method is called (which could be multiple times).

Although dynamic prerequisites sound attractive, I think you should take advantage of the built-in facilities for pre-building dependencies first to see if they fit your bill.

For example, you could say:

```
file "dependencies.rake" => SRC_FILES do
  open("dependencies.rake", "w") do |deps|
    determine_dependencies_and_write_to(deps)
  end
end

import "dependencies.rake"
```

When the import line is executed, Rake will check for a file task to build the dependency file, run the task and import the dependencies once before any of the other tasks are executed.  This allows the dependencies.rake file to define the prerequisites once before any task that needs it is run.

Furthermore, since the dependencies are stored in a file, they don't have to be recalculated as long as the source files don't change (and if they do change, the dependency file will be automatically regenerated).

You can even use dependency files written in Makefile format and use utilities like makedepend to generate.  Rake is quite happy to read makefile dependency files.

You can read about the import command at https://github.com/jimweirich/rake/blob/master/doc/rakefile.rdoc

Give that a try and see if it solves your problems.

damphyr commented:
I'm already doing something like that by using task invocation within a task:

``` ruby
task :foo do
  some_task_with_dynamic_deps=calculate_task_with_dependencies()
  Taks[some_task_with_dynamic_deps].invoke
end
```

which allows me to control caching of dependencies.

My current argument is readability. I have quite a large system and while I know the trick, people (and with people I mean my backup and "holiday guarantor") have difficulties following the dependency chain.

``` ruby
task :foo => calculate_dependencies
```

is much more readable and easier to trace(and I can rig a caching mechanism similar to import nicely encapsulated in calculate_dependencies ) .

Also, I have lots and lots of tasks. In the beginning I was determining the dependency chain before everything else (much like for the import)*. The delay in rake's startup time was frankly embarrassing (being on WIndows makes it painful already).
While statistically insignificant compared to the build tasks (10-12s vs. 10-20min) it made rake -T painful and the perception was that the system was extremely slow. 
Then I started checking the arguments to figure out if the task required us to scan for dependencies. This helped but the code became more and more convoluted the more functionality we added to the system and it also required constant fiddling with the logic that decided if the dependency generation executes or not. It was, to say the least, ugly and fragile.

With this patch I can break up the calculation logically and yet explicitly show on the first level what the task depends on

``` ruby
task :foo => [:code_dependencies,:build_dependencies]
```

With my current solution this is buried within calculate_task_with_dependencies. 

*There are some system specifics that in my case make caching (with import or without) less effective: about a third of the code is generated from a DSL. The DSL Generator (due to some design issues with Xtext 1.0) touches a lot of files even if it doesn't need to. This results in dependency calculation running more often than required.

jimweirich commented:
Although I'm kinda attracted to the idea of dynamic dependencies, what puts me off the idea is that adding a dependency may or may not do what you expect.  If the task you add the dependency to is already finished, it will have no effect.  I know that in the particular style you use this is generally not a problem, but there seems to be a great deal of potential abuse and failure, and I'm the one that will have to respond to all the "dynamic prerequisites don't work for me" bug reports in the future.

Import was supposed to be a "controlled" dynamic dependency management technique to address your issues, but you are saying that doesn't work for you.  Maybe we can come up with something that is a little less open-ended for me but still works for you.

damphyr commented:
I have to say I expected this. It's not as if rake is some niche library, so a certain conservatism in adding/expanding features is recommended practice I'd say. 
What I considered doing (partly because of the delay as well) is create a gem to monkey-patch the functionality. So then it becomes an opt-in for whoever wants to use it and any caveats can be described in the gem documentation.

I must say that using the 

``` ruby
task :foo do
  t=create_task()
  Rake::Task.invoke(t)
end
```

idiom works very well, the context of dependency generation is very focused, parameterization is easy (either through task parameters or environment variables) and responsiveness is excellent (we're down to the usual require-on-Windows sickness).
Like I said the only real reason would be visibility of dependencies and I would concede that it does not merit the tradeoff with this abuse potential.
